Support multiple reports per device in usb_hid

dhalbert opened the pull request:
This partly implements #4868. The main purpose of this PR is to update the `usb_hid` API in preparation for 7.0.0.
- Require that the report ID the user wants is in the report descriptor. It will not be changed and will not be assigned dynamically.
- Due to the above, the default devices now use these report IDs: keyboard: 1, mouse: 2, consumer control: 3.
- Remove the `Device()` constructor argument that is the report ID index.
- Change the `Device` constructor to take multiple report IDs and IN and OUT report sizes.
- Change the `.last_received_report` API for OUT reports to `get_last_received_report(report_id)`, to allow fetching from multiple report id's. The property still works for single report ID's, but will probably be removed in 8.0.0.
- Cleaned up the doc a bit.

- Also fixed a copy-pasta bug in `argcheck.c`.

Not done yet:
- Bidirectional feature reports (this may or may not require some TinyUSB changes).
- Boot devices (will be an upward compatible addition).

Tested with MacroPad hotkeys project, with a Metro M4, and with a Consumer Control volume control project.

dhalbert commented:
Wait a bit on this -- I am testing a change to the keyboard report descriptor to fix #5076.

dhalbert commented:
> Wait a bit on this -- I am testing a change to the keyboard report descriptor to fix #5076.

This problem may not be what I thought it was, so fine to review as is.

tannewt commented on shared-module/usb_hid/__init__.c:
```diff
@@ -260,24 +243,32 @@ void usb_hid_gc_collect(void) {
+
+        // Collect all the OUT report buffers for this device.
+        for (size_t id_idx = 0; id_idx < hid_devices[device_idx].num_report_ids; id_idx++) {
+            gc_collect_ptr(hid_devices[id_idx].out_report_buffers[id_idx]);
```
Collect the in report buffers too.

dhalbert replied:
Whoops, yes, thanks. I had originally removed the IN buffers, but realized I still needed them.

tannewt commented on shared-bindings/usb_hid/Device.c:
```diff
@@ -81,82 +89,144 @@ STATIC mp_obj_t usb_hid_device_make_new(const mp_obj_type_t *type, size_t n_args
+    enum { ARG_buf, ARG_report_id };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_buf, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_report_id, MP_ARG_OBJ, {.u_obj = mp_const_none} },
```
Why not just default this to -1 instead of None?

dhalbert replied:
I did think about doing that, but I want the user-facing API to be `None`, for clarity: `None` means no report id supplied. If the user does supply the parameter (instead of defaulting), it should still work the same way.

tannewt reviewed (changes requested):
Will we be able to make this a smooth transition from 6 to 7? I think most of the these APIs are below the HID library that we can update.

dhalbert commented:
> Will we be able to make this a smooth transition from 6 to 7? I think most of the these APIs are below the HID library that we can update.

The constructor `usb_hid.Device()` didn't exist before 7, so that's not an issue. I will have to change the Learn Guide.
The `.last_received_report` already existed, so I am adding `get_last_received_report(report_id)`, and then will drop `.last_received_report` in 8.

dhalbert commented:
38f4281 addresses the review comments, and also:

- The default KEYBOARD report descriptor had a signed/unsigned error, and also could have allowed more keycodes. So I revised it, using the very vanilla descriptor from a very plain extremely common commercial keyboard, modifying it only have 5 LED's instead of 3, and added a report ID. This might fix #5076, but even if not, it's a more correct report descriptor.
- Removed some byte-counting in the descriptor comments that is no longer necessary.

dhalbert commented:
Latest commit is safer about trying to write to report buffers, and addresses an issue noticed in #5020. I could not reproduce #5020 but this code is a good idea anyway.

tannewt reviewed (approved):
Looks good to me! Thanks for the tweaks.
